classes/robot/crawler.php

```diff
@@ -549,53 +551,58 @@ public function process_queue($verbose = false) {
             $recentcourses = $this->get_recentcourses();
         }
 
-        // Iterate through the queue until we find an item that is a recent course, or the time runs out.
+        // Iterate through the queue.
         $cronstart = time();
         $cronstop = $cronstart + $config->maxcrontime;
-        $hasmore = true;
         $hastime = true;
-        while ($hasmore && $hastime) {
-            // Grab the first item from the queue.
-            $node = $DB->get_record_sql('SELECT *
+
+        $timeout = 0; // Lock this resource with a zero second timeout.
+        $locktype = 'tool_crawler_process_queue';
+
+        // Get an instance of the currently configured lock_factory.
+        $lockfactory = \core\lock\lock_config::get_lock_factory($locktype);
+
+        // While we are not exceeding the maxcron time, and the queue is not empty.
+        while ($hastime) {
+            if (empty($nodes)) {
+                // Grab a list of items from the front of the queue
+                $nodes = $DB->get_records_sql('SELECT *
                                          FROM {tool_crawler_url}
                                         WHERE lastcrawled IS NULL
                                            OR lastcrawled < needscrawl
                                      ORDER BY priority DESC, needscrawl ASC, id ASC
-                                        LIMIT 1
-                                    ');
+                                        LIMIT :limitadhocworkers
```

Comment:
I think limitadhocworkers is a misleading name for this

Reply:
also limits should be done using the 3rd and 4th arguments to get_records_sql to be cross db

Reply:
Not sure about a great name, I've updated the name to availablequeueitem and added a more detailed comment above.